---
Title: Go 1.14でシステムコールがEINTRエラーを返すようになった
Category:
- Go
Date: 2020-02-29T16:27:27+09:00
URL: https://blog.lufia.org/entry/2020/02/29/162727
EditURL: https://blog.hatena.ne.jp/lufiabb/lufia.hateblo.jp/atom/entry/26006613527875969
---

Go 1.13までのゴルーチンの切り替えは、チャネルの送受信やシステムコール呼び出し、関数呼び出し前にコンパイラが暗黙的に挿入する処理などによって行われていました。そのため、上記の切り替わり操作を全く行わないループなどがあれば、そのゴルーチンがずっと実行されます。

```go
func loop() {
	// この辺りにコンパイラがコード挿入している

	for {
		// 切り替わり処理が行われないので無限に実行される
	}
}
```

この結果、*$GOMAXPROCS*が1の場合はプログラムが停止します。コンパイラが挿入するコードは、インライン展開された場合や`go:nosplit`ディレクティブが記述された場合には行われないので、関数呼び出しをしていてもゴルーチンが切り替わらない場合はあります。

* [ゴルーチンの切り替えと関数のスタックチェック](https://qiita.com/tobi-c/items/41e6c786c858a513f67e)
* [go doc compile/Compiler Directives](https://golang.org/cmd/compile/#hdr-Compiler_Directives)

Goのスケジューラについてはこの辺りが詳しいです。

* [GoConで発表してきたのでついでにruntime以下の知識をまとめていく](https://niconegoto.hatenadiary.jp/entry/2017/04/11/092810)
* [Golangのスケジューラあたりの話](https://qiita.com/takc923/items/de68671ea889d8df6904)

これ自体はGoを学び始めた人がよく引っかかるものですね。

## 何が問題なのか

このように、Goでは特定の場所でしかゴルーチンの切り替えが行われないため、[Non-cooperative goroutine preemption](https://github.com/golang/proposal/blob/master/design/24543-non-cooperative-preemption.md)というプロポーザル((issueは[golang/go#24543](https://github.com/golang/go/issues/24543)です))によると、GCのStopTheWorld時間が長くなったりするそうです。

Go 1.14では、上記のようなコードでも切り替えられるように、ゴルーチンの切り替えが*SA_RESTART*フラグ付きの*SIGURG*によっても行われるようになりました。ざっとコードを眺めた雰囲気だと、*SIGURG*はGCの処理で送っているみたいですね。この結果、[Go 1.14のリリースノート](https://golang.org/doc/go1.14)には

> A consequence of the implementation of preemption is that on Unix systems, including Linux and macOS systems, programs built with Go 1.14 will receive more signals than programs built with earlier releases. This means that programs that use packages like [syscall](https://golang.org/pkg/syscall/) or [golang.org/x/sys/unix](https://godoc.org/golang.org/x/sys/unix) will see more slow system calls fail with EINTR errors.

と書かれています。システムコール実行中にシグナルを受け取ると、システムコールは*EINTR*を返す場合があるので、[syscall](https://golang.org/pkg/syscall/)や[golang.org/x/sys/unix](https://pkg.go.dev/golang.org/x/sys/unix)を直接使っている場合は適切なエラーハンドリングしましょう、とのことです。*os*や*net*など、他のパッケージを使っている場合もたぶん同じで、Go 1.13まではエラーにならなかったシステムコールが*EINTR*エラーになるパターンが増えます。

* [errno(3)](https://linuxjm.osdn.jp/html/LDP_man-pages/man3/errno.3.html)

## どう対応するべきか

Linuxの場合((他のOSでも概ね同じだと思いますが))はシステムコールや呼び出し時のオプションによって、シグナルで中断した場合の動作は以下の2通りに分かれます。

- シグナルを処理した後で再開される
- *EINTR*エラーを返す

どのシステムコールがどんな条件で*EINTR*を返すのかは、Linuxマニュアルに詳しく書かれていました。

* [signal(7)](https://linuxjm.osdn.jp/html/LDP_man-pages/man7/signal.7.html)
	* シグナルハンドラーによるシステムコールやライブラリ関数への割り込み
	* 一時停止シグナルによるシステムコールやライブラリ関数への割り込み

自動でカーネルが再開してくれる場合は、当然ですが今までと同じように動作するため対応不要です。Goランタイムが送る*SIGURG*には*SA_RESTART*フラグが付いているので、ほとんどは大丈夫そう。*EINTR*を返すことがあるシステムコールの場合は、そのままエラーを返すかリトライするかを決める必要があるでしょう。とはいえ*syscall.EINTR*を直接使ってしまうと、他のOSに移植する際にとても困ります((*syscall*が提供している型を公開メソッドで直接参照するのは良くないと思います))。Goの*syscall.Errno*は*Temporary*メソッドを実装していて、*EINTR*の場合は*true*を返すので、代わりにこれを使うと良いと思います。

```go
type temporaryer interface {
	Temporary() bool
}

_, err := r.Read(buf)
if err != nil {
	if e, ok := err.(temporaryer); ok {
		fmt.Println(e.Temporary())
	}
}
```

ただし、*close(2)*の場合、システムコールから戻った時点でファイルディスクリプタは無効になっており、同じファイルディスクリプタが他のファイルに割り当てられる可能性があるため、リトライしてはいけないようです。

[https://twitter.com/gachacomplete/status/1232478292529168384:embed]

* [closeがEINTRを返したらどうするべきか](https://togetter.com/li/583503)

## おまけ

[Goでファイルの存在確認](https://blog.lufia.org/entry/2019/05/07/234521)でも似たようなことを書きましたが、ファイルの書き込みを行う場合、`defer`を使って、

```go
w, err := os.OpenFile(file, os.O_WRONLY|os.O_APPEND, 0660)
if err != nil {
	return err
}
defer w.Close()

_, err := w.Write(data)
return err
```

みたいに書いてしまうと、*close(2)*のエラーが拾えません。その結果、書き込みが失敗しているのに成功扱いしてしまいます。読み込みだけの場合はこれで問題ありませんが、書き込みをしたファイルの*Close*は、OSが持っているバッファをディスクに書き込もうとしてエラーになる可能性があるので、きちんとエラーを拾いましょう。

### 2020-03-02追記

ファイルの場合、*os.File.Sync* でバッファをフラッシュしておいて、*Close*ではエラーを発生させないようにするのがベストプラクティスのようです。

[https://twitter.com/gachacomplete/status/1233685816032911360?s=20:embed]
